File: java-sdk/src/main/java/com/cleanster/sdk/client/CleansterConfig.java
package com.cleanster.sdk.client;

/**
 * Configuration for the Cleanster SDK client.
 * Use the builder to construct instances.
 */
public class CleansterConfig {

    public static final String SANDBOX_BASE_URL =
            "https://partner-sandbox-dot-official-tidyio-project.ue.r.appspot.com/public";
    public static final String PRODUCTION_BASE_URL =
            "https://partner-dot-official-tidyio-project.ue.r.appspot.com/public";

    private final String baseUrl;
    private final String accessKey;
    private final int connectTimeoutSeconds;
    private final int readTimeoutSeconds;
    private final int writeTimeoutSeconds;

    private CleansterConfig(Builder builder) {
        this.baseUrl = builder.baseUrl;
        this.accessKey = builder.accessKey;
        this.connectTimeoutSeconds = builder.connectTimeoutSeconds;
        this.readTimeoutSeconds = builder.readTimeoutSeconds;
        this.writeTimeoutSeconds = builder.writeTimeoutSeconds;
    }

    public String getBaseUrl() {
        return baseUrl;
    }

    public String getAccessKey() {
        return accessKey;
    }

    public int getConnectTimeoutSeconds() {
        return connectTimeoutSeconds;
    }

    public int getReadTimeoutSeconds() {
        return readTimeoutSeconds;
    }

    public int getWriteTimeoutSeconds() {
        return writeTimeoutSeconds;
    }

    /**
     * Create a new builder for sandbox environment.
     *
     * @param accessKey Your Cleanster partner API access key
     */
    public static Builder sandboxBuilder(String accessKey) {
        return new Builder(accessKey).baseUrl(SANDBOX_BASE_URL);
    }

    /**
     * Create a new builder for production environment.
     *
     * @param accessKey Your Cleanster partner API access key
     */
    public static Builder productionBuilder(String accessKey) {
        return new Builder(accessKey).baseUrl(PRODUCTION_BASE_URL);
    }

    public static class Builder {
        private String baseUrl = SANDBOX_BASE_URL;
        private final String accessKey;
        private int connectTimeoutSeconds = 30;
        private int readTimeoutSeconds = 60;
        private int writeTimeoutSeconds = 60;

        public Builder(String accessKey) {
            if (accessKey == null || accessKey.isBlank()) {
                throw new IllegalArgumentException("accessKey must not be null or blank");
            }
            this.accessKey = accessKey;
        }

        public Builder baseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        public Builder connectTimeoutSeconds(int seconds) {
            this.connectTimeoutSeconds = seconds;
            return this;
        }

        public Builder readTimeoutSeconds(int seconds) {
            this.readTimeoutSeconds = seconds;
            return this;
        }

        public Builder writeTimeoutSeconds(int seconds) {
            this.writeTimeoutSeconds = seconds;
            return this;
        }

        public CleansterConfig build() {
            return new CleansterConfig(this);
        }
    }
}

File: csharp-sdk/src/Cleanster/CleansterConfig.cs
namespace Cleanster;

/// <summary>
/// Configuration for <see cref="CleansterClient"/>.
/// Create via <see cref="Sandbox"/> or <see cref="Production"/> factory methods,
/// or construct directly for custom base URLs or timeouts.
/// </summary>
public sealed class CleansterConfig
{
    /// <summary>Base URL for the sandbox environment (development/testing — no real charges).</summary>
    public const string SandboxBaseUrl =
        "https://partner-sandbox-dot-official-tidyio-project.ue.r.appspot.com/public";

    /// <summary>Base URL for the production environment (live traffic — real cleaners, real charges).</summary>
    public const string ProductionBaseUrl =
        "https://partner-dot-official-tidyio-project.ue.r.appspot.com/public";

    /// <summary>Default HTTP request timeout (30 seconds).</summary>
    public static readonly TimeSpan DefaultTimeout = TimeSpan.FromSeconds(30);

    /// <summary>Your partner access key — sent as the "access-key" header on every request.</summary>
    public string AccessKey { get; }

    /// <summary>API base URL.</summary>
    public string BaseUrl { get; }

    /// <summary>HTTP request timeout.</summary>
    public TimeSpan Timeout { get; }

    /// <summary>
    /// Create a config with explicit values.
    /// </summary>
    /// <param name="accessKey">Your partner access key (must not be blank).</param>
    /// <param name="baseUrl">API base URL (must not be blank).</param>
    /// <param name="timeout">HTTP timeout (defaults to <see cref="DefaultTimeout"/>).</param>
    /// <exception cref="ArgumentException">Thrown when <paramref name="accessKey"/> or <paramref name="baseUrl"/> is blank.</exception>
    public CleansterConfig(string accessKey, string baseUrl, TimeSpan? timeout = null)
    {
        if (string.IsNullOrWhiteSpace(accessKey))
            throw new ArgumentException("Cleanster: accessKey must not be blank.", nameof(accessKey));
        if (string.IsNullOrWhiteSpace(baseUrl))
            throw new ArgumentException("Cleanster: baseUrl must not be blank.", nameof(baseUrl));

        AccessKey = accessKey;
        BaseUrl   = baseUrl;
        Timeout   = timeout ?? DefaultTimeout;
    }

    /// <summary>Returns a config pre-configured for the sandbox environment.</summary>
    public static CleansterConfig Sandbox(string accessKey)
        => new(accessKey, SandboxBaseUrl);

    /// <summary>Returns a config pre-configured for the production environment.</summary>
    public static CleansterConfig Production(string accessKey)
        => new(accessKey, ProductionBaseUrl);
}
